### PR TITLE
additional labels / affinity rules calculated by machine driver

### PR DESCRIPTION
**Pull request by ibrokethecloud**
Related issue: https://github.com/harvester/harvester/issues/4588

When a VM set is launched from the Harvester UI, and anti-affinity rules are specified, then Harvester enhances the VM definitions to add additional labels in `vm.spec.metadata.labels`

```
      labels:
        harvesterhci.io/vmName: ubuntu-vmset-01
        harvesterhci.io/vmNamePrefix: ubuntu-vmset
```

The additional `vmNamePrefix` label is used during defining `podAffinity` or `podAntiAffinity`

```
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchLabels:
                  harvesterhci.io/vmNamePrefix: ubuntu-vmset
              topologyKey: kubernetes.io/hostname
            weight: 1
          requiredDuringSchedulingIgnoredDuringExecution:
          - topologyKey: topology.kubernetes.io/zone
      domain:
```

Rancher machine driver does not pass this information when provisioning nodes.

We can however evaluate the rancher `machineset` name from the VM name as it follows the convention `clusterName-poolName-generatedID`

The PR leverages this logic to add an additional label to VM's 
```
  labels:
    harvesterhci.io/creator: docker-machine-driver-harvester
    harvesterhci.io/machineSetName: affinity2-pool1-0d403cc4
    harvesterhci.io/vmName: affinity2-pool1-0d403cc4-4qzs7
```

This additional label can then be used to generated more refined anti affinity rules

```
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchLabels:
                  harvesterhci.io/machineSetName: affinity2-pool1-0d403cc4
              topologyKey: kubernetes.io/hostname
            weight: 1
          requiredDuringSchedulingIgnoredDuringExecution:
          - topologyKey: topology.kubernetes.io/zone
```

### CODE REVIEW

**Review by connorkuehl — APPROVED** [review]

**Comment by berkitamas** [comment]
Thank you for the fix! I have some questions, regarding the default affinity and anti-affinity terms. If I understand the changes correctly, it will add a default affinity term for the hypervisors in case there are other affinity terms defined, and the same logic applies for anti-affinity.
I don't see the logical connection of _any_ affinity term and hostname affinity term based on pool. If I set up an affinity term between the VMs, I don't necessarily want the pool members to be on the same hypervisor. 
Also, including the generatedID would make it hard (or even impossible) to set up specific anti-affinity terms because the generatedID is only known after the pool is reconfigured (including the affinity term). It would be a race condition since we need to include the generatedID before we actually know it.

**Comment by FrankYang0529** [comment]
> Also, including the generatedID would make it hard (or even impossible) to set up specific anti-affinity terms because the generatedID is only known after the pool is reconfigured (including the affinity term). It would be a race condition since we need to include the generatedID before we actually know it.

Hi @berkitamas, if I understand correctly, this change set a new label without generatedID and add it to anti-affinity, so users don't need to manually set it.

```go
		machineSetSplit := strings.Split(d.MachineName, "-")
		machineSetName := strings.Join(machineSetSplit[:len(machineSetSplit)-1], "-")
		vmBuilder = vmBuilder.Labels(map[string]string{poolNameLabelKey: machineSetName})
```

@ibrokethecloud, not sure whether the example in your description is wrong. There is generatedID on it.

File: harvester/create.go [hunk 1]
```diff
@@ -46,6 +48,31 @@ func (d *Driver) Create() error {
 		if err = json.Unmarshal([]byte(d.VMAffinity), &affinity); err != nil {
 			return err
 		}
+		//VM naming convention is of form: clusterName-poolName-generatedString
+		//we can reverse split this to identify unique machinesets name, to label nodes
+		//with this unique machine set. This can then be used for populating affinity rules
+		machineSetSplit := strings.Split(d.MachineName, "-")
+		machineSetName := strings.Join(machineSetSplit[:len(machineSetSplit)-1], "-")
```
Comment (FrankYang0529) [inline harvester/create.go:55]: It looks like the machine name is `clusterName-poolName-generatedString-randomString`. In my cluster, the machine is `rke2-3-pool1-e0b8749d-g9jln`, `rke2-3-pool1-e0b8749d-r96bc`, and `rke2-3-pool1-e0b8749d-wrbf8`. Do we want to use `strings.Join(machineSetSplit[:len(machineSetSplit)-2], "-")`?

Comment (ibrokethecloud) [reply harvester/create.go:55]: we could use either one, but i need to check if it is feasible to have same pool names

Comment (FrankYang0529) [reply harvester/create.go:55]: Yeah, there may have conflicts if users create clusters with the same name in different namespaces. Probably, we can use `<cluster-namespace>-<cluster-name>-<pool-name>`.

Comment (ibrokethecloud) [reply harvester/create.go:55]: yeah that might work.. I will make the change.

Comment (ibrokethecloud) [reply harvester/create.go]: PR now includes additional logic to drop generatedID from pool name and includes namespace.

**Review by FrankYang0529 — APPROVED** [review]
LGTM. Thanks.

**Review by bk201 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**